Release 2.22.0 (#8536)
timescale/timescaledb: CHANGELOG.md

**Highlighted features in TimescaleDB v2.22.0**
* Sparse indexes on compressed hypertables can now be explicitly configured via `ALTER TABLE` rather than relying only on internal heuristics. Users can define indexes on multiple columns to improve query performance for their specific workloads.
* [Tech Preview] Continuous aggregates now support the `timescaledb.invalidate_using` option, enabling invalidations to be collected either via triggers on the hypertable or directly from WAL using logical decoding. Aggregates inherit the hypertable’s method if none is specified.  
* UUIDv7 compression and vectorization are now supported. The compression algorithm leverages the timestamp portion for delta-delta compression while storing the random portion separately. The vectorized equality/inequality filters with bulk decompression deliver ~2× faster query performance. The feature is disabled by default (`timescaledb.enable_uuid_compression`) to simplify the downgrading experience, and will be enabled out of the box in the next minor release.

> * Hypertables can now be partitioned by UUIDv7 columns, using their embedded timestamp for time-based chunking. Non-v7 UUIDs are rejected on insert.  

Reviewer: ```suggestion
* Hypertables can now be partitioned by UUIDv7 columns, using their embedded timestamp for time-based chunking. Alongside the addition of utility functions for simplified generation of UUIDv7 values or extracting timestamps, e.g. `uuid_timestamp()` will return a Postgres timestamp value from the UUIDv7.
```

`psql` with the `-X` flag to prevent any `.psqlrc` commands from
accidentally triggering the load of a previous DB version.**

## 2.22.0 (2025-09-02)

This release contains performance improvements and bug fixes since the 2.21.3 release. We recommend that you upgrade at the next available opportunity.

**Highlighted features in TimescaleDB v2.22.0**
* [#8247](https://github.com/timescale/timescaledb/pull/8247) Add configurable alter settings for sparse indexes

> **Removal of the hypercore access method**

Reviewer 2: ```suggestion
**Sunsetted the hypercore table access method**
```

Not sure, but we say sunsetted in other places

> We made the decision to deprecate hypercore access method (TAM) with the 2.21.0 release. It was an experiment, which did not show the signals we hoped for and is removed with this release. Upgrading to 2.22.0 and higher will be blocked if TAM is still in use. Since TAM’s inception in [2.18.0](https://github.com/timescale/timescaledb/releases/tag/2.18.0), we learned that btrees were not the right architecture. The recent advancements in the columnstore—such as more performant backfilling, SkipScan, adding check constraints, and faster point queries—put the [columnstore](https://www.timescale.com/blog/hypercore-a-hybrid-row-storage-engine-for-real-time-analytics) close to or on par with TAM without the storage from the additional index. We apologize for the inconvenience this action potentially causes and are here to assist you during the migration process.

Reviewer 2: ```suggestion
We made the decision to deprecate the hypercore table access method (TAM) with the 2.21.0 release. Hypercore TAM was an experiment and it did not show the performance improvements we hoped for. It is removed with this release. Upgrades to 2.22.0 and higher are blocked if TAM is still in use. Since TAM’s inception in [2.18.0](https://github.com/timescale/timescaledb/releases/tag/2.18.0), we learned that btrees were not the right architecture. Recent advancements in the columnstore, such as more performant backfilling, SkipScan, adding check constraints, and faster point queries, put the [columnstore](https://www.timescale.com/blog/hypercore-a-hybrid-row-storage-engine-for-real-time-analytics) close to or on par with TAM without needing to store an additional index. We apologize for the inconvenience this action potentially causes and are here to assist you during the migration process.
```

Migration path

```
do $$
declare
   relid regclass;
begin
   for relid in
       select cl.oid from pg_class cl
       join pg_am am on (am.oid = cl.relam)
       where am.amname = 'hypercore'
   loop
       raise notice 'converting % to heap', relid::regclass;
       execute format('alter table %s set access method heap', relid);
   end loop;
end
$$;
```

**Features**
* [#8247](https://github.com/timescale/timescaledb/pull/8247) Add configurable alter settings for sparse indexes
* [#8306](https://github.com/timescale/timescaledb/pull/8306) Add option for invalidation collection using WAL for continuous aggregates

> * [#8340](https://github.com/timescale/timescaledb/pull/8340) Improve selectivity estimates for sparse minmax indexes, so that an index scan on compressed table is chosen more often when it's beneficial.

Reviewer 2: ```suggestion
* [#8340](https://github.com/timescale/timescaledb/pull/8340) Improve selectivity estimates for sparse minmax indexes, so that an index scan on a table in the columnstore is chosen more often when it's beneficial.
```

* [#8360](https://github.com/timescale/timescaledb/pull/8360) Continuous aggregate multi-hypertable invalidation processing
* [#8364](https://github.com/timescale/timescaledb/pull/8364) Remove hypercore table access method

> * [#8371](https://github.com/timescale/timescaledb/pull/8371) Show available timescaledb ALTER options when encountering unsupported options

Reviewer 2: ```suggestion
* [#8371](https://github.com/timescale/timescaledb/pull/8371) Show available timescaledb `ALTER` options when encountering unsupported options
```

> * [#8376](https://github.com/timescale/timescaledb/pull/8376) Change DecompressChunk custom node name to ColumnarScan

Reviewer 2: ```suggestion
* [#8376](https://github.com/timescale/timescaledb/pull/8376) Change `DecompressChunk` custom node name to `ColumnarScan`
```

* [#8385](https://github.com/timescale/timescaledb/pull/8385) UUID v7 functions for testing pre PG18
* [#8393](https://github.com/timescale/timescaledb/pull/8393) Add specialized compression for UUIDs. Best suited for UUID v7, but still works with other UUID versions. This is experimental at the moment and backward compatibility is not guaranteed.
* [#8398](https://github.com/timescale/timescaledb/pull/8398) Set default compression settings at compress time

> * [#8401](https://github.com/timescale/timescaledb/pull/8401) Support ALTER TABLE RESET for compression settings

Reviewer 2: ```suggestion
* [#8401](https://github.com/timescale/timescaledb/pull/8401) Support `ALTER TABLE RESET` for compression settings
```

* [#8414](https://github.com/timescale/timescaledb/pull/8414) Vectorised filtering of UUID Eq and Ne filters, plus bulk decompression of UUIDs

> * [#8424](https://github.com/timescale/timescaledb/pull/8424) Block downgrade when orderby setting is NULL

Reviewer 2: ```suggestion
* [#8424](https://github.com/timescale/timescaledb/pull/8424) Block downgrade when orderby setting is `NULL`
```

> `psql` with the `-X` flag to prevent any `.psqlrc` commands from

Reviewer 2: I know this is not part of the ticket, but is there any chance we can update the intro sentence to say:

This page lists all the latest features and updates to TimescaleDB. When you use `psql` to update your database, use  `-X` flag and prevent any `.psqlrc` commands from accidentally triggering the load of a previous DB version.

Reviewer: @billy-the-fish please make a suggestion in the file
